lib/livebook/storage/ets.ex

```diff
@@ -98,13 +110,90 @@ defmodule Livebook.Storage.Ets do
 
     :ets.insert(table, attributes)
 
-    {:reply, :ok, state}
+    {:reply, :ok, schedule_persist(state)}
   end
 
   @impl GenServer
   def handle_call({:delete, namespace, entity_id}, _from, %{table: table} = state) do
     :ets.delete(table, {namespace, entity_id})
 
-    {:reply, :ok, state}
+    {:reply, :ok, schedule_persist(state)}
+  end
+
+  @impl GenServer
+  def handle_info(:persist, state) do
+    state = %{state | pending_persist: nil}
+
+    {:noreply, save_to_file(state)}
+  end
+
+  defp schedule_persist(%{pending_persist: ref} = state) do
+    unless is_nil(ref) do
+      Process.cancel_timer(ref)
+    end
+
+    ref = Process.send_after(self(), :persist, @persistence_timeout)
+
+    %{state | pending_persist: ref}
+  end
+
+  defp default_secret() do
+    Livebook.Config.secret!("LIVEBOOK_SECRET_KEY_BASE")
+  end
+
+  defp default_file_path() do
+    # NOTE: should we add the application port to distinguish between multiple instances?
+    Path.join([:filename.basedir(:user_cache, "livebook"), "livebook.conf"])
+  end
+
+  defp load_from_file(%{file_path: file_path, secret: secret} = state)
+       when is_nil(file_path) or is_nil(secret) do
+    state
+  end
+
+  defp load_from_file(%{table: table, file_path: file_path, secret: secret} = state) do
+    # NOTE: should we silently ignore a case where the file does not exist yet?
+    if File.exists?(file_path) do
+      file_path
+      |> load()
+      |> decrypt(secret)
+      |> :erlang.binary_to_term()
+      |> then(&:ets.insert(table, &1))
+    end
+
+    state
```

Comment:
Nit: since we are not updating the state, it would be clearer to return `:ok` :)